pallas-txbuilder/src/transaction/model.rs

```diff
@@ -0,0 +1,712 @@
+pub struct Datum {
+    pub kind: DatumKind,
+    pub bytes: DatumBytes,
+}
+
+#[derive(PartialEq, Eq, Hash, Debug)]
+pub enum RedeemerPurpose {
+    Spend(Input),
+    Mint(PolicyId),
+    // Reward TODO
+    // Cert TODO
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct ExUnits {
+    pub mem: u32,
+    pub steps: u64,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct Redeemers(HashMap<RedeemerPurpose, (Bytes, Option<ExUnits>)>);
+
+impl Deref for Redeemers {
+    type Target = HashMap<RedeemerPurpose, (Bytes, Option<ExUnits>)>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Redeemers {
+    pub fn new() -> Self {
+        Redeemers(HashMap::new())
+    }
+
+    pub fn from_map(map: HashMap<RedeemerPurpose, (Bytes, Option<ExUnits>)>) -> Self {
+        Self(map)
+    }
+}
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct Address(pub PallasAddress);
+
+impl Deref for Address {
+    type Target = PallasAddress;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl From<PallasAddress> for Address {
+    fn from(value: PallasAddress) -> Self {
+        Self(value)
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+#[serde(rename_all = "snake_case")]
+pub enum BuilderEra {
+    Babbage,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+pub struct BuiltTransaction {
+    pub version: String,
+    pub era: BuilderEra,
+    pub status: TransactionStatus,
+    pub tx_hash: TxHash,
+    pub tx_bytes: Bytes,
+    pub signatures: Option<HashMap<PublicKey, Signature>>,
+}
+
+impl BuiltTransaction {
+    pub fn sign(mut self, secret_key: ed25519::SecretKey) -> Result<Self, TxBuilderError> {
+        let pubkey: [u8; 32] = secret_key
+            .public_key()
+            .as_ref()
+            .try_into()
+            .map_err(|_| TxBuilderError::MalformedKey)?;
+
+        let signature: [u8; 64] = secret_key.sign(self.tx_hash.0).as_ref().try_into().unwrap();
+
+        match self.era {
+            BuilderEra::Babbage => {
+                let mut new_sigs = self.signatures.unwrap_or_default();
+
+                new_sigs.insert(Bytes32(pubkey), Bytes64(signature));
+
+                self.signatures = Some(new_sigs);
+
+                // TODO: chance for serialisation round trip issues?
+                let mut tx = babbage::Tx::decode_fragment(&self.tx_hash.0)
+                    .map_err(|_| TxBuilderError::CorruptedTxBytes)?;
+
+                let mut vkey_witnesses = tx.transaction_witness_set.vkeywitness.unwrap_or_default();
+
+                vkey_witnesses.push(babbage::VKeyWitness {
+                    vkey: Vec::from(pubkey.as_ref()).into(),
+                    signature: Vec::from(signature.as_ref()).into(),
+                });
+
+                tx.transaction_witness_set.vkeywitness = Some(vkey_witnesses);
+
+                self.tx_bytes = tx.encode_fragment().unwrap().into();
```

Comment:
Doesn't this mean the hash field should be updated as well?

Reply:
Transaction hash is only of the transaction body, here we only modify the transaction witnesses

Reply:
oh duh, ok